### DISCUSSION TITLE
/hyndsight/coherent-population-forecasting/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#195] [Announcements]
Coherent population forecasting using R 
 https://robjhyndman.com/hyndsight/coherent-population-forecasting/

**Comment by robjhyndman** [comment]
```header
Originally posted by buggyfunbunny on 2014-07-25 02:52:44
````
I am always puzzled by attempts to forecast the human condition.

This is from the link, &quot;Why are some\...&quot;:

\- What is nor­mally assumed is that the way the envi­ron­ment is chang­ing
will con­tinue into the future.

Yet that assumption, with respect to the human condition, is clearly
false. Casual civilians, at least here in the USofA, sop up the
assertion that &quot;people are living so much longer than when Social
Security was created, so we can\'t have Social Security any more!!!&quot;
And similar. Of course, that\'s hogwash. Yes, life expectancy at birth
has increased by a couple of decades since 1935. But life expectancy at
65 has gone up by a few years; controlling for wealth (which is to say,
access to quality health care), not even so much. Longevity at birth
increase happened due to huge decrease in infant/child mortality,
antibiotics, and vaccines. That ship has sailed. One might argue, I\'ve
not data to hand, that increase in longevity for adults is largely due
to decrease in smoking and other affronts. Certainly, cancer and cardio-
survival hasn\'t really increased all that much (taken for a population
as a whole) over the last 50 years.

Life expectancy change, esp. increase, is a direct result of human
intervention in the human condition \-- new findings in science and
medicine; it\'s not the end result of the workings of Mother Nature,
Natural Law, or God. It\'s not exogenous and it\'s not the result of
some continuous process. Specific new findings led to specific new
therapies. From 1900 forward, we\'ve learned enough to know that there
aren\'t many unknown unknowns left in science (at human scale;
sub-atomic and cosmologic excepted). Increase in longevity, for the 99%
anyway, is about played out. Given the proliferation of the Western
Diet, i.e. obesity and diabetes, and likely restriction of access to
quality health care as the result of increasing wealth inequality,
regression is likely.

Historical data, in this context, shows us only the net result of
science up to today. Tomorrow will be the net result of new science and
behavioral changes (less smoking, for example). Whether there\'s any new
science that can have much more than a year or two of additional
longevity? Not something I would bet on.

Certainly, I wouldn\'t bet that the trajectory of science/medicine on
longevity of the last 50 years can be sustained over the next 50.

**Comment by robjhyndman** [comment]
```header
Originally posted by Syazreen on 2014-12-10 22:45:38
````
Hi Prof,\
The pop.sim function allows only (&quot;fmforecast2&quot;) for mortf class. That
mean we have to coherently model males and females. If I want to model
males and females independently (using fdm instead of coherentfdm), can
I use pop.sim to project population? Thanks

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-12-10 23:10:04
````
No. You would need to write your own functions if you want to do that.

**Reply by robjhyndman** [reply]
```header
Originally posted by Syazreen on 2014-12-10 23:23:56
````
Ok. Thanks

**Comment by robjhyndman** [comment]
```header
Originally posted by Syazreen on 2015-09-01 03:57:59
````
Hi Prof,\
I wonder why the older age (60+) prediction intervals seem does not
exist. It does happen to Malaysia population too.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-09-01 07:54:02
````
It\'s very narrow because the population is extremely easy to predict
for high ages.

**Comment by robjhyndman** [comment]
```header
Originally posted by Qi Cui on 2016-12-08 01:15:47
````
Hi Professor,\
When you forecast fertility, the original fertility data is from 1920 to
2002, so if do the forecast, the forecasted year should start from 2003.
Also if fix the starting year - 1950, the end year of based data between
mortality and fertility are different, so how to forecast the net
migrants starting from 2010?

Many thanks,\
Qi

**Comment by robjhyndman** [comment]
```header
Originally posted by Ture on 2017-03-10 23:59:38
````
This is really interesting. Could you please give an example where you
use textfiles and generate the same or similar results? I find it hard
to replicate the example with my own data (in txt format) but would love
to succeed.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-03-12 21:10:24
````
Create demogdata objects, then proceed in the same way. See
?demography::demogdata

**Comment by robjhyndman** [comment]
```header
Originally posted by Werner Grundlingh on 2017-06-14 23:23:20
````
Smoothing the data for aus.fert does not work. The following causes an
error:

[ausfert.sm](http://ausfert.sm &quot;ausfert.sm&quot;)
\&lt;- smooth.demogdata(aus.fert, obs.var=&quot;theoretical&quot;)

Could this be a problem with that component of the data set?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-15 04:33:29
````
Works for me. Please provide a minimal reproducible example as described
at
[https://robjhyndman.com/hyn\...](https://robjhyndman.com/hyndsight/minimal-reproducible-examples/ &quot;https://robjhyndman.com/hyndsight/minimal-reproducible-examples/&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted by Werner Grundlingh on 2017-06-15 05:38:30
````
My sequence included:

1\) Downloaded ausdata.RData from the link provided
([https://robjhyndman.com/Rfi\...](https://robjhyndman.com/Rfiles/ausdata.RData) &quot;https://robjhyndman.com/Rfiles/ausdata.RData)&quot;).

2\) Loaded ausdata.RData into RStudio.

3\) Executed the following sequence:

library(demography)\
aus.mig \&lt;- netmigration(aus.mort, aus.fert, mfratio = 1.05)\
[ausmort.sm](http://ausmort.sm &quot;ausmort.sm&quot;)
\&lt;- smooth.demogdata(aus.mort, obs.var=&quot;theoretical&quot;)\
[ausfert.sm](http://ausfert.sm &quot;ausfert.sm&quot;)
\&lt;- smooth.demogdata(aus.fert, obs.var=&quot;theoretical&quot;)\
[ausmig.sm](http://ausmig.sm &quot;ausmig.sm&quot;) \&lt;-
smooth.demogdata(aus.mig)

\[Console:
[http://imgur.com/a/1eBji)](http://imgur.com/a/1eBji) &quot;http://imgur.com/a/1eBji)&quot;)

4\) Error when trying to smooth
[ausfert.sm](http://ausfert.sm &quot;ausfert.sm&quot;):

\&gt; [ausfert.sm](http://ausfert.sm &quot;ausfert.sm&quot;)
\&lt;- smooth.demogdata(aus.fert, obs.var=&quot;theoretical&quot;)\
Error in .local(x, nrow, ncol, eps, \...) : object \'f_coocsr\' not
found

\-\-\--

For what it\'s worth, devtools::session_info() yields

Session info
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--\
setting value\
version R version 3.4.0 (2017-04-21)\
system x86_64, mingw32\
ui RStudio (1.0.143)\
language (EN)\
collate English_Canada.1252\
tz America/Los_Angeles\
date 2017-06-14

Packages
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--\
package \* version date source\
base \* 3.4.0 2017-04-21 local\
cluster 2.0.6 2017-03-10 CRAN (R 3.4.0)\
cobs 1.3-3 2017-03-31 CRAN (R 3.3.3)\
codetools 0.2-15 2016-10-05 CRAN (R 3.4.0)\
colorspace 1.3-2 2016-12-14 CRAN (R 3.3.2)\
compiler 3.4.0 2017-04-21 local\
datasets \* 3.4.0 2017-04-21 local\
demography \* 1.20 2017-04-25 CRAN (R 3.4.0)\
devtools 1.13.2 2017-06-02 CRAN (R 3.4.0)\
digest 0.6.12 2017-01-27 CRAN (R 3.3.2)\
fda 2.4.4 2014-12-16 CRAN (R 3.3.3)\
FNN 1.1 2013-07-31 CRAN (R 3.3.3)\
forecast \* 8.0 2017-02-23 CRAN (R 3.3.2)\
fracdiff 1.4-2 2012-12-02 CRAN (R 3.2.3)\
ftsa 4.8 2017-04-25 CRAN (R 3.4.0)\
ggplot2 2.2.1 2016-12-30 CRAN (R 3.3.2)\
graphics \* 3.4.0 2017-04-21 local\
grDevices \* 3.4.0 2017-04-21 local\
grid 3.4.0 2017-04-21 local\
gtable 0.2.0 2016-02-26 CRAN (R 3.2.3)\
hdrcde 3.1 2013-10-19 CRAN (R 3.3.3)\
htmltools 0.3.6 2017-04-28 CRAN (R 3.3.3)\
htmlwidgets 0.8 2016-11-09 CRAN (R 3.3.2)\
httpuv 1.3.3 2015-08-04 CRAN (R 3.2.3)\
jsonlite 1.5 2017-06-01 CRAN (R 3.3.3)\
KernSmooth 2.23-15 2015-06-29 CRAN (R 3.4.0)\
knitr 1.16 2017-05-18 CRAN (R 3.3.3)\
ks 1.10.6 2017-03-20 CRAN (R 3.3.3)\
lattice 0.20-35 2017-03-25 CRAN (R 3.4.0)\
lazyeval 0.2.0 2016-06-12 CRAN (R 3.2.5)\
lmtest 0.9-35 2017-02-11 CRAN (R 3.3.2)\
magrittr 1.5 2014-11-22 CRAN (R 3.2.3)\
MASS 7.3-47 2017-02-26 CRAN (R 3.4.0)\
Matrix 1.2-9 2017-03-14 CRAN (R 3.4.0)\
MatrixModels 0.4-1 2015-08-22 CRAN (R 3.3.3)\
memoise 1.1.0 2017-04-21 CRAN (R 3.3.3)\
methods \* 3.4.0 2017-04-21 local\
mgcv 1.8-17 2017-02-08 CRAN (R 3.4.0)\
mime 0.5 2016-07-07 CRAN (R 3.2.5)\
misc3d 0.8-4 2013-01-25 CRAN (R 3.3.3)\
multicool 0.1-10 2016-11-25 CRAN (R 3.3.3)\
munsell 0.4.3 2016-02-13 CRAN (R 3.2.3)\
mvtnorm 1.0-6 2017-03-02 CRAN (R 3.3.2)\
nlme 3.1-131 2017-02-06 CRAN (R 3.4.0)\
nnet 7.3-12 2016-02-02 CRAN (R 3.4.0)\
parallel 3.4.0 2017-04-21 local\
pcaPP 1.9-61 2016-10-11 CRAN (R 3.3.2)\
plyr 1.8.4 2016-06-08 CRAN (R 3.3.2)\
quadprog 1.5-5 2013-04-17 CRAN (R 3.2.3)\
quantmod 0.4-9 2017-05-29 CRAN (R 3.3.3)\
quantreg 5.33 2017-04-18 CRAN (R 3.3.3)\
R6 2.2.1 2017-05-10 CRAN (R 3.3.3)\
rainbow 3.4 2016-01-03 CRAN (R 3.3.3)\
Rcpp 0.12.11 2017-05-22 CRAN (R 3.3.3)\
rgl 0.98.1 2017-03-08 CRAN (R 3.3.3)\
rlang 0.1.1 2017-05-18 CRAN (R 3.3.3)\
scales 0.4.1 2016-11-09 CRAN (R 3.3.2)\
sde 2.0.15 2016-04-13 CRAN (R 3.3.3)\
shiny 1.0.3 2017-04-26 CRAN (R 3.3.3)\
SparseM 1.77 2017-04-23 CRAN (R 3.3.3)\
splines 3.4.0 2017-04-21 local\
stats \* 3.4.0 2017-04-21 local\
stats4 3.4.0 2017-04-21 local\
tibble 1.3.3 2017-05-28 CRAN (R 3.3.3)\
timeDate 3012.100 2015-01-23 CRAN (R 3.2.3)\
tools 3.4.0 2017-04-21 local\
tseries 0.10-41 2017-05-29 CRAN (R 3.3.3)\
TTR 0.23-1 2016-03-21 CRAN (R 3.3.3)\
utils \* 3.4.0 2017-04-21 local\
withr 1.0.2 2016-06-20 CRAN (R 3.2.5)\
xtable 1.8-2 2016-02-05 CRAN (R 3.2.3)\
xts 0.9-7 2014-01-02 CRAN (R 3.3.3)\
zoo 1.8-0 2017-04-12 CRAN (R 3.3.3)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-15 07:12:43
````
I can\'t reproduce the problem. But I notice that you have lots of
packages installed under different versions of R, which is almost
certainly going to lead to difficulties. Please update all packages
using R3.4.0 and try again.

**Reply by robjhyndman** [reply]
```header
Originally posted by Werner Grundlingh on 2017-06-21 20:42:18
````
Thanks, this fixed the problem. Do you know of a way to update all
packages using the latest version of R? I\'m running on Windows 10 using
RStudio, if that matters.

**Comment by robjhyndman** [comment]
```header
Originally posted by Alex on 2018-02-15 02:08:16
````
Fantastic post as always RH!

Quick question though - i\'m having lots of trouble creating a demogdata
object when not using data from the HMD. I\'ve traced through the
docuementation/github trying to debug my code but to no avail.

Would you be able to show the code you used to coerce the ABS fertility
data into a demogdata object from the blog post above? I don\'t suppose
it\'s hiding on your github profile somewhere?

Alex

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-02-15 06:03:33
````
Every format is different, so examples are unlikely to be helpful. I
suggest you use demogdata() directly.

**Reply by robjhyndman** [reply]
```header
Originally posted by Alex on 2018-02-15 06:54:37
````
Perhaps i wasn\'t clear in my question, i have been using demogdata()
directly. But i\'m hitting this error when constructing my data and i
can\'t figure out why this test is in place.

From demogdata() function:\
n \&lt;- ncol(data)\
if(length(years) != n)\
stop(&quot;Number of years doesn\'t match data&quot;)

I don\'t understand why the vector of years should be of corresponding
length to the number of columns?

For example,

Data (data.frame):\
Year, Age, Female, Male, Total\
1921, 0, \... \... \...\
1921, 1, \... \... \...\
1921, 2, \... \... \...\
1921, 3, \... \... \...\
\...\
2016, 109, \... \... \...\
2016, 110+, \... \... \...

Years (vector):\
1921,\
1921,\
1921,\
1921,\
\...\
2014,\
2014.

ncol(df) = 5\
length(years) = 10434

They don\'t equal each other \... ERROR!

Thanks for the help.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-02-15 07:20:33
````
The matrix of data has years in the columns and ages in the rows. So the
number of columns should match the number of years. Your data frame is
not in the right format for this function.

**Comment by robjhyndman** [comment]
```header
Originally posted by Sher Akbar on 2019-01-05 12:30:34
````
hi. prof.\
how can i get the data in discrete format may be in excel or in r
format\...\
as the data you have in pkg: demography is in functional format..

**Comment by robjhyndman** [comment]
```header
Originally posted by Carlos Ahumada on 2019-04-10 13:08:47
````
Hi Prof!

I have been trying to create a demogdata object with my own population
.txt file but cannot find a way to do it since the demogdata and
read.demogdata functions require an additional matrix with fertility or
mortality rates. This would be important for the pop.sim function
because I cannot select the *firstyearpop* argument that it requires. I
would really appreciate any feedback/support from you.

Thanks in advance!

**Comment by PUTOPIS** [comment]
Hi,

In my opinion, no need to divide with 1000, for example here:
 B <- colSums(fert$pop$female * fert$rate$female/1000)
  Bf <- B * 1/(1 + 1.05)
  Bm <- B * 1.05/(1 + 1.05)

Best regards,
Lidija

**Reply by robjhyndman** [reply]
The fertility rates are stored as births per 1000 women.

**Reply by PUTOPIS** [reply]
Thank you for answer!

If data (rates) are correctly imported with read.demogdata (with scale), the fertility rates are stored as births per 1 woman.

But, doesn't matter! Now projections seem good.

Regards from Zagreb (Croatia)!
